peps/pep-0705.rst

```diff
@@ -363,6 +365,35 @@ Notes:
     b: B = { "x": 3 }
     c: C = a | b  # Accepted by type checker: "x" will always come from b
 
+Minimum support
```

Comment:
I think the minimum support should be to "enforce the behaviors indicated in typeshed". Anything beyond that requires a type checker to override the typeshed stubs, which is something I'm very reluctant to do in pyright unless there's really strong reason to do otherwise. I've made exceptions in only a couple of places so far, and those have been ongoing maintenance burdens.

I've received one or two requests for improving the `|` operator for TypedDict, but that's nowhere near enough to compel me to add special-case behavior. I'd need to receive dozens of more requests before considering it. As for `copy` and `deepcopy`, I've never had a request for improving these for TypedDict operands, so I'm not be inclined to add such support.

I'm curious why you've focused on `|`, `copy` and `deepcopy`. Have you personally experienced issues with these? Or are you trying to solve a problem that is completely theoretical here?

The motivation section doesn't mention `|`, `copy` or `deepcopy` as being problems for TypedDict. If these are going to be a focus on this PEP, I think they should feature prominently in the motivation and/or rationale sections. Right now, these seem to come out of nowhere.

What you've outlined below seems like a good sketch of how a type checker _could_ override the standard typeshed stubs if desired, but I'd prefer that it be completely optional and left up to each type checker to decide whether to do so. I'm not sure this even belongs in a PEP.

Reply:
Re: merge. This was more important when other_keys was in the spec. I think I'm happy to defer.

Re: copy/deepcopy: I think the typeshed stubs will be functionally deficient for typeddicts with read-only entries, and users are likely to hit it and complain. Without these changes, AFAIK there's no way to copy a readonly typeddict and change an item in it without an unsafe cast? This is a new problem so it's not a surprise if it's not come up before.

I may be missing subtleties here though. For instance, are you only talking about `copy.copy` and not `TypedDict.copy`?

Reply:
> This is a new problem so it's not a surprise if it's not come up before.

Then I'd prefer to address it if and when pyright users run into the issue, not be required to do it speculatively because the PEP mandates it.

I don't think there's any precedent for a typing PEP to mandate that a type checker provide custom logic that overrides typeshed stubs. Not even the PEP 557 (dataclass) or PEP 589 (TypedDict) makes any such statement, although PEP 589 does come close (in that it specifies cases where the `get` and `clear` methods in [this section](https://peps.python.org/pep-0589/#supported-and-unsupported-operations).

Perhaps it would be best if PEP 705 stuck to defining the meaning of `ReadOnly` and how this affects type consistency rules. Individual type checkers can then choose how to implement support for it.

> are you only talking about `copy.copy` and not `TypedDict.copy`?

I'm talking about both, but I'm more concerned about `copy.copy` and `copy.deepcopy`. The `TypedDict.copy` method doesn't bother me as much because TypedDict classes and methods are already special-cased by type checkers. The `copy.copy` and `copy.deepcopy` methods are very general functions. I suspect they are rarely used in conjunction with TypedDict, especially when there are straightforward alternatives including `TypedDict.copy` and `{**td}`.

Reply:
> I  don't think there's any precedent for a typing PEP to mandate that a type checker provide custom logic that overrides typeshed stubs.

I've been missing the obvious: you're telling me I need to target this kind of change at the typeshed stubs, not the type checkers.

> straightforward alternatives including...{**td}.

I just checked and mypy out-of-the box appears to support *all* the rules I specified if you write:

```python
c: C = { **a, **b }
```

Yet the functionally identical code with `|` is rejected. The same issue holds for copy.

I am surprised that the status quo is apparently to move users away from idiomatic and readable modern Python onto the spread operator, but you're quite right, this does make all of this technically unnecessary.

Reply:
For what it's worth, there are some active efforts to improve `|` support in mypy: https://github.com/python/mypy/pull/16249